csrc/contiguity.cpp

```diff
@@ -87,292 +90,401 @@ bool OrderedIdInformation::checkExclusivelyConsumesAllocs(IterDomain* id) {
+    const auto& in_alloc_ids = findAllocIDs(split->in())->second;
 
-  // Update map for outputs
-  // Remove inputs from the active_ids_ and insert the output ID
-  active_ids_[in_pos] = split->outer();
-  active_ids_.insert(active_ids_.begin() + in_pos + 1, split->inner());
+    // Update map for outputs
+    // Remove inputs from the active_ids_ and insert the output ID
+    active_ids_[in_pos] = split->outer();
+    active_ids_.insert(active_ids_.begin() + in_pos + 1, split->inner());
 
-  // The outputs are ordered as long as the input is ordered.
-  if (in_ordered) {
-    consistently_ordered_ids_.emplace(split->outer());
-    consistently_ordered_ids_.emplace(split->inner());
-  }
+    // The outputs are ordered as long as the input is ordered.
+    if (in_ordered) {
+      consistently_ordered_ids_.emplace(split->outer());
+      consistently_ordered_ids_.emplace(split->inner());
+    }
+
+    // Update the alloc_id entry for the outputs.
+    id_to_alloc_ids_[split->outer()] = in_alloc_ids;
+    id_to_alloc_ids_[split->inner()] = in_alloc_ids;
+  } else { // Direction::Backward
+    // Backward analysis assumes use of IdModel
+    if (!using_id_graph_) {
+      return;
+    }
+
+    if (!isActiveId(split->outer()) || !isActiveId(split->inner())) {
+      return;
+    }
+
+    const auto outer_pos = getActiveIdPos(split->outer());
+    const auto inner_pos = getActiveIdPos(split->inner());
 
-  // Update the alloc_id entry for the outputs.
-  id_to_alloc_ids_[split->outer()] = in_alloc_ids;
-  id_to_alloc_ids_[split->inner()] = in_alloc_ids;
+    const bool outer_ordered = isConsistentlyOrdered(split->outer());
+    const bool inner_ordered = isConsistentlyOrdered(split->inner());
+
+    // Get allocation ids of the two inputs
+    const auto& outer_alloc_ids = findAllocIDs(split->outer())->second;
+    const auto& inner_alloc_ids = findAllocIDs(split->inner())->second;
+
+    // The output is ordered as long as the inputs are ordered and outer is
+    // directly to the left of inner. The forward analysis of Merge considers
+    // the case where the inner position is not directly next to the outer
+    // position because there may be broadcast. However, broadcast IDs should
+    // not be included in analyzed allocation IDs when IdModel is used.
+    bool out_ordered =
+        outer_ordered && inner_ordered && (inner_pos == outer_pos + 1);
+
+    if (out_ordered) {
+      consistently_ordered_ids_.emplace(split->in());
+    }
+
+    active_ids_[outer_pos] = split->in();
+
+    if (out_ordered) {
+      active_ids_.erase(active_ids_.begin() + inner_pos);
+    } else {
+      active_ids_[inner_pos] = nullptr;
+    }
+
+    VectorOfUniqueEntries<IterDomain*> alloc_ids = inner_alloc_ids;
+    alloc_ids.pushBack(outer_alloc_ids);
+    id_to_alloc_ids_[split->in()] = alloc_ids;
+
+    if (checkExclusivelyConsumesAllocs(split->in())) {
+      exclusively_consumes_allocs_.emplace(split->in());
+    }
+  }
```

Comment:
This part is new. Based on the forward merge case. Simplified as use of IdModel is assumed.